distributed/tests/test_nanny.py

```diff
@@ -472,21 +474,39 @@ async def test_nanny_closed_by_keyboard_interrupt(ucx_loop, protocol):
             assert "remove-worker" in str(s.events)
 
 
-class StartException(Exception):
-    pass
-
-
 class BrokenWorker(worker.Worker):
-    async def start(self):
-        raise StartException("broken")
+    def __init__(self, startup_attempts, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.startup_attempts = startup_attempts
+
+    async def start_unsafe(self):
+        with self.startup_attempts.get_lock():
+            self.startup_attempts.value += 1
+        raise ValueError("broken")
 
 
 @gen_cluster(nthreads=[])
 async def test_worker_start_exception(s):
-    # make sure this raises the right Exception:
-    with raises_with_cause(RuntimeError, None, StartException, None):
-        async with Nanny(s.address, worker_class=BrokenWorker) as n:
-            pass
+    startup_attempts = mp.Value("i", 0)
+    nanny = Nanny(
+        s.address, worker_class=BrokenWorker, startup_attempts=startup_attempts
+    )
+    with captured_logger(logger="distributed.nanny", level=logging.WARNING) as logs:
+        with raises_with_cause(
+            RuntimeError,
+            "Nanny failed to start",
+            RuntimeError,
+            "BrokenWorker failed to start",
+        ):
+            async with nanny:
+                pass
+    assert nanny.status == Status.failed
+    # ^ NOTE: `Nanny.close` sets it to `closed`, then `Server.start._close_on_failure` sets it to `failed`
+    assert nanny.process is None
+    assert startup_attempts.value == 1
+    assert "Restarting worker" not in logs.getvalue()
+    # Avoid excessive spewing. (It's also printed once extra within the subprocess, which is okay.)
+    assert logs.getvalue().count("ValueError: broken") == 1, logs.getvalue()
```

Comment:
@fjetter I didn't remove the logging capture yet (though I could 00e277edcb26152dd0a1cebc4193332d886ccd14), because I liked having this assertion. I noticed that previously, the exception was being logged multiple times from the nanny side, which would make things even noisier and harder to decipher for users.

I did add a multiprocessing Value to count the startup attempts, but I'm not really sure this is necessary or not.

Reply:
I've removed the multiprocessing Value. It seems unnecessary. And in one test, restarting the worker repeatedly segfaulted the Python interpreter. I have no evidence for this, but I feel like maybe some broken state in a shared Value could cause that.